### PR TITLE
feat: support querying partition tables through select in

### PR DESCRIPTION
**Pull request by ZuLiangWang**
# Which issue does this PR close?

Closes #

# Rationale for this change
We implemented table partition in the last release, but querying partitioned tables through `select in` will lead to the failure to identify the correct partitions, resulting in full table scanning. We need to support select in to improve query performance.

# What changes are included in this PR?
* Implement parsing of `select in` in `filterExtractor`

# Are there any user-facing changes?
None.

# How does this change test
Pass unit test.

### CODE REVIEW

File: table_engine/src/partition/rule/df_adapter/extractor.rs [hunk 1]
```diff
@@ -67,6 +67,30 @@ impl FilterExtractor for KeyExtractor {
                     }
                     _ => None,
                 },
+                Expr::InList {
+                    expr: _,
```
Comment (Rachelint) [inline table_engine/src/partition/rule/df_adapter/extractor.rs:71]: We should check the expr, we only process  Expr::Column(col).

File: table_engine/src/partition/rule/df_adapter/extractor.rs [hunk 2]
```diff
@@ -67,6 +67,30 @@ impl FilterExtractor for KeyExtractor {
...
+                    list,
+                    negated,
+                } => {
+                    // The 'not in' syntax is not currently supported.
+                    if negated {
+                        continue;
+                    }
```
Comment (Rachelint) [inline table_engine/src/partition/rule/df_adapter/extractor.rs:78]: We can use tuple match to check here and above, for example:
`if let (Expr::Column(col), true) = (*expr, negated) `

File: table_engine/src/partition/rule/df_adapter/extractor.rs [hunk 3]
```diff
@@ -67,6 +67,30 @@ impl FilterExtractor for KeyExtractor {
...
+                    let mut v = Vec::with_capacity(list.len());
+                    for e in list {
```
Comment (Rachelint) [inline table_engine/src/partition/rule/df_adapter/extractor.rs:80]: Maybe e -> entry

File: table_engine/src/partition/rule/df_adapter/extractor.rs [hunk 4]
```diff
@@ -67,6 +67,30 @@ impl FilterExtractor for KeyExtractor {
...
+                        if let Expr::Literal(val) = e {
+                            let datum_opt = Datum::from_scalar_value(&val);
+                            if let Some(d) = datum_opt {
```
Comment (Rachelint) [inline table_engine/src/partition/rule/df_adapter/extractor.rs:83]: d -> datum

File: table_engine/src/partition/rule/df_adapter/extractor.rs [hunk 5]
```diff
@@ -67,6 +67,30 @@ impl FilterExtractor for KeyExtractor {
...
+                                v.push(d)
+                            }
+                        }
+                    }
+                    let column_name = columns_in_filter.iter().next().unwrap();
```
Comment (Rachelint) [inline table_engine/src/partition/rule/df_adapter/extractor.rs:88]: We should use the column name which get from `expr`.

File: table_engine/src/partition/rule/df_adapter/extractor.rs [hunk 6]
```diff
@@ -67,6 +67,30 @@ impl FilterExtractor for KeyExtractor {
...
+                Expr::InList {
+                    expr: _,
+                    list,
+                    negated,
+                } => {
+                    // The 'not in' syntax is not currently supported.
+                    if negated {
+                        continue;
+                    }
+                    let mut v = Vec::with_capacity(list.len());
```
Comment (Rachelint) [inline table_engine/src/partition/rule/df_adapter/extractor.rs:79]: v -> datums

File: table_engine/src/partition/rule/df_adapter/extractor.rs [hunk 7]
```diff
@@ -67,6 +67,30 @@ impl FilterExtractor for KeyExtractor {
...
+                            if let Some(d) = datum_opt {
+                                v.push(d)
+                            }
+                        }
+                    }
+                    let column_name = columns_in_filter.iter().next().unwrap();
+                    Some(PartitionFilter::new(
+                        column_name.clone(),
+                        PartitionCondition::In(v),
+                    ))
```
Comment (Rachelint) [inline table_engine/src/partition/rule/df_adapter/extractor.rs:92]: Check datums, if it is empty, we should return None.

File: table_engine/src/partition/rule/df_adapter/extractor.rs [hunk 8]
```diff
@@ -107,4 +131,24 @@ mod tests {
...
+    fn test_key_extractor_select_in() {
+        let extractor = KeyExtractor;
+
+        let columns = vec!["col1".to_string()];
+        let accepted_expr = col("col1").in_list(
+            vec![
+                Literal(ScalarValue::Int32(Some(42))),
+                Literal(ScalarValue::Int32(Some(38))),
+            ],
+            false,
```
Comment (Rachelint) [inline table_engine/src/partition/rule/df_adapter/extractor.rs:145]: Negated should be tested.

File: table_engine/src/partition/rule/df_adapter/extractor.rs [hunk 9]
```diff
@@ -107,4 +131,24 @@ mod tests {
         let partition_filter = extractor.extract(&[rejected_expr], &columns);
         assert!(partition_filter.is_empty());
     }
+
+    #[test]
+    fn test_key_extractor_select_in() {
```
Comment (Rachelint) [inline table_engine/src/partition/rule/df_adapter/extractor.rs:136]: Maybe select_in -> in_list_fitler.

File: table_engine/src/partition/rule/df_adapter/extractor.rs [hunk 10]
```diff
@@ -63,7 +63,35 @@ impl FilterExtractor for KeyExtractor {
...
+                    }
+                    _ => None,
+                },
+                Expr::InList {
+                    expr,
+                    list,
+                    negated,
+                } => match (*expr, list, negated) {
+                    // The 'not in' syntax is not currently supported.
+                    (Expr::Column(col), list, false) => {
```
Comment (Rachelint) [inline table_engine/src/partition/rule/df_adapter/extractor.rs:78]: `list` is unneccessary to match.

File: table_engine/src/partition/rule/df_adapter/extractor.rs [hunk 11]
```diff
@@ -63,7 +63,35 @@ impl FilterExtractor for KeyExtractor {
...
+                        })
+                    }
+                    _ => None,
+                },
+                Expr::InList {
+                    expr,
+                    list,
+                    negated,
+                } => match (*expr, list, negated) {
+                    // The 'not in' syntax is not currently supported.
```
Comment (Rachelint) [inline table_engine/src/partition/rule/df_adapter/extractor.rs:77]: Can not support actually, maybe this comment can be removed.

File: table_engine/src/partition/rule/df_adapter/extractor.rs [hunk 12]
```diff
@@ -63,7 +63,35 @@ impl FilterExtractor for KeyExtractor {
...
+                                }
+                            }
+                        }
+                        match datums.is_empty() {
+                            false => Some(PartitionFilter::new(
+                                col.name,
+                                PartitionCondition::In(datums),
+                            )),
+                            true => None,
+                        }
```
Comment (Rachelint) [inline table_engine/src/partition/rule/df_adapter/extractor.rs:94]: For bool, usually use `if else`.

File: table_engine/src/partition/rule/df_adapter/extractor.rs [hunk 13]
```diff
@@ -107,4 +135,40 @@ mod tests {
...
+        let columns = vec!["col1".to_string()];
+        let accepted_expr = col("col1").in_list(
+            vec![
+                Literal(ScalarValue::Int32(Some(42))),
+                Literal(ScalarValue::Int32(Some(38))),
+            ],
+            true,
+        );
+        let partition_filter = extractor.extract(&[accepted_expr], &columns);
+        assert_eq!(partition_filter.len(), 0);
```
Comment (Rachelint) [inline table_engine/src/partition/rule/df_adapter/extractor.rs:172]: Usually check `is_empty`.

File: table_engine/src/partition/rule/df_adapter/extractor.rs (hunk 10 above)
Comment (ZuLiangWang) [reply table_engine/src/partition/rule/df_adapter/extractor.rs:78]: `list` is moved in  `match (*expr, list, negated)`, so we must declare it again.

Comment (Rachelint) [reply table_engine/src/partition/rule/df_adapter/extractor.rs:77]: It seems strange?
How about:
```suggestion
{
      if  let (Expr::Column(col), false) = (Expr::Column(col), false) {
         ...
      } else {
           None
      }
 }
```

**Comment by Rachelint** [comment]
LGTM

**Review by Rachelint — APPROVED** [review]
LGTM

**Pull request merged**